### PR TITLE
Add `Table` update code path

### PR DESCRIPTION
**Pull request by A-Hilaly**
Issues:

- https://github.com/aws-controllers-k8s/community/issues/1000
- https://github.com/aws-controllers-k8s/community/issues/1029
- https://github.com/aws-controllers-k8s/community/issues/1596

Description of changes:
- Add `Table` custom update code
- Add `Table` custom preCompare delta code (to properly compare
`GlobalSecondaryIndexes` ,`AttributeDefinitions` and `KeySchemas`)
- Add unit tests for custom compare code for `GlobalSecondaryIndexes`
- Make `KeySchema` an immutable field.
- Gradually update `GlobalSecondaryIndexes` when multiple ones are
update/removed or added.
- Add custom code to properly populate `BillingMode` and
`SSESpecification` and `TableClass`
- Add e2e tests for each new support field update + one test to ensure
  that multi-field updates at once can be properly handled

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/table/hooks.go [hunk 1]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
+	// - Modify the provisioned throughput settings of the table.
+	// - Enable or disable DynamoDB Streams on the table.
+	// - Remove a global secondary index from the table.
+	// - Create a new global secondary index on the table. After the index begins
+	//   backfilling, you can use UpdateTable to perform other operations.
+	//
+	// UpdateTable is an asynchronous operation; while it is executing, the table
+	// status changes from ACTIVE to UPDATING. While it is UPDATING, you cannot
+	// issue another UpdateTable request. When the table returns to the ACTIVE state,
+	// the UpdateTable operation is complete.
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:155]: Ah this is the same as EKS clusters. For that, I threw [a requeue error](https://github.com/aws-controllers-k8s/eks-controller/blob/main/pkg/resource/cluster/hook.go#L59) after making an update, giving the system a 15 second time to wait for the EKS service to register the update, start the async update and update the cluster status description. Do we need to do the same thing here, or does the Dynamo table immediately start updating and update the status? That is, if I describe the table instantly after calling `UpdateTable` will it show as `UPDATING`?

File: pkg/resource/table/hooks.go [hunk 2]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
-		if err != nil {
-			return err
+		if r.ko.Spec.ProvisionedThroughput.WriteCapacityUnits != nil {
+			input.ProvisionedThroughput.WriteCapacityUnits = aws.Int64(*r.ko.Spec.ProvisionedThroughput.WriteCapacityUnits)
+		} else {
+			input.ProvisionedThroughput.WriteCapacityUnits = aws.Int64(0)
 		}
+	} else {
+		input.ProvisionedThroughput.ReadCapacityUnits = aws.Int64(0)
+		input.ProvisionedThroughput.WriteCapacityUnits = aws.Int64(0)
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:218]: Oh jeez please tell me this isn't possible? lol

File: pkg/resource/table/hooks.go [hunk 3]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
-// resourceTagsFromSDKTags transforms a *svcsdk.Tag array to a *v1alpha1.Tag array.
-func resourceTagsFromSDKTags(svcTags []*svcsdk.Tag) []*v1alpha1.Tag {
-	tags := make([]*v1alpha1.Tag, len(svcTags))
-	for i := range svcTags {
-		tags[i] = &v1alpha1.Tag{
-			Key:   svcTags[i].Key,
-			Value: svcTags[i].Value,
-		}
+// syncTable updates a given table billing mode, stream specification
+// or SEE specification.
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:230]: nit:
```suggestion
// or SSE specification.
```

File: pkg/resource/table/hooks.go [hunk 4]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
+			input.StreamSpecification = &svcsdk.StreamSpecification{
+				StreamEnabled: aws.Bool(false),
 			}
 		}
-		removed = append(removed, aElement.Key)
 	}
-	for _, bElement := range b {
-		if !ackutil.InStrings(*bElement.Key, visitedIndexes) {
-			added = append(added, bElement)
+	if delta.DifferentAt("Spec.SEESpecification") {
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:286]:
```suggestion
	if delta.DifferentAt("Spec.SSESpecification") {
```

File: pkg/resource/table/hooks.go [hunk 5]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
+	// - Create a new global secondary index on the table. After the index begins
+	//   backfilling, you can use UpdateTable to perform other operations.
+	//
+	// UpdateTable is an asynchronous operation; while it is executing, the table
+	// status changes from ACTIVE to UPDATING. While it is UPDATING, you cannot
+	// issue another UpdateTable request. When the table returns to the ACTIVE state,
+	// the UpdateTable operation is complete.
+
+	if delta.DifferentAt("Spec.BillingMode") ||
+		delta.DifferentAt("Spec.SEESpecification") {
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:158]:
```suggestion
		delta.DifferentAt("Spec.SSESpecification") {
```

File: pkg/resource/table/hooks.go [hunk 6]
```diff
@@ -140,139 +140,164 @@ func (rm *resourceManager) customUpdateTable(
...
-			&svcsdk.UntagResourceInput{
-				ResourceArn: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
-				TagKeys:     removed,
-			},
-		)
-		rm.metrics.RecordAPICall("GET", "UntagResource", err)
-		if err != nil {
-			return err
+	if r.ko.Spec.ProvisionedThroughput != nil {
+		if r.ko.Spec.ProvisionedThroughput.ReadCapacityUnits != nil {
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks.go:205]: Should this be using `newSDKProvisionedThroughtput`?

File: pkg/resource/table/hooks_tags.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+	if len(removed) > 0 {
+		_, err = rm.sdkapi.UntagResourceWithContext(
+			ctx,
+			&svcsdk.UntagResourceInput{
+				ResourceArn: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
+				TagKeys:     removed,
+			},
+		)
+		rm.metrics.RecordAPICall("GET", "UntagResource", err)
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks_tags.go:58]:
```suggestion
		rm.metrics.RecordAPICall("UPDATE", "UntagResource", err)
```

File: pkg/resource/table/hooks_tags.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+	if len(added) > 0 {
+		_, err = rm.sdkapi.TagResourceWithContext(
+			ctx,
+			&svcsdk.TagResourceInput{
+				ResourceArn: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
+				Tags:        sdkTagsFromResourceTags(added),
+			},
+		)
+		rm.metrics.RecordAPICall("GET", "UntagResource", err)
```
Comment (RedbackThomson) [inline pkg/resource/table/hooks_tags.go:72]:
```suggestion
		rm.metrics.RecordAPICall("UPDATE", "UntagResource", err)
```

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
Looks pretty ready. Just found some typos in your diffs, and I have a question regarding re-queueing because of the nature of asynchronous updates.

File: pkg/resource/table/hooks.go [hunk 9]
```diff
@@ -140,139 +150,166 @@ func (rm *resourceManager) customUpdateTable(
...
+				StreamEnabled: aws.Bool(false),
 			}
 		}
-		removed = append(removed, aElement.Key)
 	}
-	for _, bElement := range b {
-		if !ackutil.InStrings(*bElement.Key, visitedIndexes) {
-			added = append(added, bElement)
+	if delta.DifferentAt("Spec.SSESpecification") {
+		if r.ko.Spec.SSESpecification.Enabled != nil {
```
Comment (jaypipes) [inline pkg/resource/table/hooks.go:299]: You need a nil guard here on `r.ko.Spec.SSESpecification`. It's causing a nil dereference fault:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x166b8ee]
goroutine 544 [running]:
github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table.(*resourceManager).newUpdateTablePayload(0xc00074def0, {0x1a6373b, 0xc}, 0xc00000e600, 0x0)
	/github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table/hooks.go:299 +0x36e
github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table.(*resourceManager).syncTable(0xc00021d100, {0x1dc6ed8, 0xc00074df20}, 0x0, 0x15bd865)
	/github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table/hooks.go:252 +0x14a
github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table.(*resourceManager).customUpdateTable(0xc00021d100, {0x1dc6ed8, 0xc00074df20}, 0xc00000e600, 0xc00000e6c8, 0xd)
	/github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table/hooks.go:169 +0x565
github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table.(*resourceManager).sdkUpdate(...)
	/github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table/sdk.go:987
github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table.(*resourceManager).Update(0xc00074def0, {0x1dc6ed8, 0xc00074df20}, {0x1df7278, 0xc00000e600}, {0x1df7278, 0xc00000e6c8}, 0xc00108a440)
	/github.com/aws-controllers-k8s/dynamodb-controller/pkg/resource/table/manager.go:148 +0x74
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).updateResource(0xc00017ca20, {0x1dc6ed8, 0xc00074df20}, {0x1df3310, 0xc00021d100}, {0x1df7278, 0xc00000e600}, {0x1df7278, 0xc00000e6c8})
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.17.0/pkg/runtime/reconciler.go:413 +0x305
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Sync(0xc00017ca20, {0x1dc6ed8, 0xc00074df20}, {0x1df3310, 0xc00021d100}, {0x1df7278, 0xc00000e600})
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.17.0/pkg/runtime/reconciler.go:240 +0x6c5
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).reconcile(0xc00017ca20, {0x1dc6ed8, 0xc00074df20}, {0x1df3310, 0xc00021d100}, {0x1df7278, 0xc00000e600})
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.17.0/pkg/runtime/reconciler.go:191 +0xe5
github.com/aws-controllers-k8s/runtime/pkg/runtime.(*resourceReconciler).Reconcile(0xc00017ca20, {0x1dc6ed8, 0xc000977f20}, {{{0xc00070f6e5, 0x1995580}, {0xc0009e84a0, 0x30}}})
	/go/pkg/mod/github.com/aws-controllers-k8s/runtime@v0.17.0/pkg/runtime/reconciler.go:172 +0x725
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0007e58c0, {0x1dc6ed8, 0xc000977e60}, {{{0xc00070f6e5, 0x1995580}, {0xc0009e84a0, 0x413834}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0007e58c0, {0x1dc6e30, 0xc0005d7dc0}, {0x18af8c0, 0xc000422880})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0007e58c0, {0x1dc6e30, 0xc0005d7dc0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:223 +0x357 
```

**Review by jaypipes — CHANGES_REQUESTED** [review]

File: pkg/resource/table/hooks.go (hunk 2 above)
Comment (A-Hilaly) [reply pkg/resource/table/hooks.go:230]: It is not lol, but i think if a user deletes these fields that what the controller would be expected to do?

File: pkg/resource/table/hooks.go (hunk 1 above)
Comment (A-Hilaly) [reply pkg/resource/table/hooks.go:165]: Gotcha! Dynamodb dosen't always instantly start updating, i added the same code path you have with EKS.

File: generator.yaml [hunk 10]
```diff
@@ -8,9 +14,25 @@ operations:
 resources:
   Table:
     fields:
+      AttributeDefinitions:
+        compare:
+          is_ignored: true
+      KeySchema:
+        compare:
+          is_immutable: true
```
[the remark this answers is not in the document]
Comment (A-Hilaly) [reply generator.yaml:22]: `is_immutable` doesn't ignore the field comparison in the generated delta function. The reason I added an `is_ignored: true` is that the Dynamodb control plane sorts the KeySchema array, and sometimes causes the controller to reconcile for ever since `reflect.DeepEqual` takes in consideration the order to elements. I added a custom `equalKeySchemaArrays` function which properly compares the Custom Resource `Spec.KeySchema` with the sorted one returned by the dynamodb control plane.

**Comment by jaypipes** [comment]
@A-Hilaly you want to rebase this and get it going again?

**Review by jaypipes — COMMENTED** [review]
@A-Hilaly let's merge https://github.com/aws-controllers-k8s/dynamodb-controller/pull/52 and rebase this please. I'd like to get this going ASAP.

**Comment by Julian-Chu** [comment]
Hi folks,

seem this PR opened for long time.
any plan to make this PR merged or break down into small new PRs?

**Comment by jaypipes** [comment]
@A-Hilaly please re-open this one.

**Comment by A-Hilaly** [comment]
Notes to reviewers, i'm debugging some failing tests in prow. Please hold on this for now.

**Comment by A-Hilaly** [comment]
@jaypipes @Julian-Chu rebased and pushed latest changes PTAL

File: test/e2e/resources/table_billing.yaml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Table used to test multiple interfering updates at once
```
Comment (RedbackThomson) [inline test/e2e/resources/table_billing.yaml:1]: I'm not a fan of these file naming conventions. The test files are related to arbitrary attribute definition names, rather than the functionality they're testing. `table_billing` could become `table_basic`, `table_forums` becomes `table_local_secondary_indexes` and `table_offices` becomes `table_global_secondary_indexes`

File: test/e2e/tests/test_table.py [hunk 12]
```diff
@@ -67,22 +89,67 @@ def forum_table():
...
+    except:
+        pass
 
 @service_marker
 @pytest.mark.canary
 class TestTable:
     def table_exists(self, table_name: str) -> bool:
         return table.get(table_name) is not None
 
+    #@pytest.mark.skipif(True, reason="temporary")
```
Comment (RedbackThomson) [inline test/e2e/tests/test_table.py:152]: Delete please :)

File: test/e2e/resources/table_billing.yaml (hunk 11 above)
Comment (A-Hilaly) [reply test/e2e/resources/table_billing.yaml:1]: Agreed :+1:

File: generator.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jaypipes) [reply generator.yaml:32]: @A-Hilaly `is_immutable` is a configuration on the FieldConfig, not the CompareConfig, so this is being ignored entirely...

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+// equalGlobalSecondaryIndexesArrays returns true if two GlobalSecondaryIndex
+// arrays are equal regardless of the order of their elements.
+func equalGlobalSecondaryIndexesArrays(
+	a []*v1alpha1.GlobalSecondaryIndex,
+	b []*v1alpha1.GlobalSecondaryIndex,
+) bool {
+	added, updated, removed := computeGlobalSecondaryIndexDelta(a, b)
+	return len(added) == 0 && len(updated) == 0 && len(removed) == 0
+}
+
```
Comment (jaypipes) [inline pkg/resource/table/hooks_global_secondary_indexes.go:79]: I don't see this function being used anywhere?

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+		input.GlobalSecondaryIndexUpdates = append(input.GlobalSecondaryIndexUpdates, update)
+		// We can only remove, update or add one GSI at once. Hence we return the update call input
+		// after we find the first removed GSI.
+		return input, gsisInQueue, nil
+	}
+	return input, gsisInQueue, nil
+}
+
+// newSDKProvisionedThroughtput builds a new *svcsdk.ProvisionedThroughput
+func newSDKProvisionedThroughtput(pt *v1alpha1.ProvisionedThroughput) *svcsdk.ProvisionedThroughput {
```
Comment (jaypipes) [inline pkg/resource/table/hooks_global_secondary_indexes.go:210]:
```suggestion
func newSDKProvisionedThroughput(pt *v1alpha1.ProvisionedThroughput) *svcsdk.ProvisionedThroughput {
```

File: pkg/resource/table/hooks_global_secondary_indexes.go [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+		}
+		input.GlobalSecondaryIndexUpdates = append(input.GlobalSecondaryIndexUpdates, update)
+		// We can only remove, update or add one GSI at once. Hence we return the update call input
+		// after we find the first removed GSI.
+		return input, gsisInQueue, nil
+	}
+	return input, gsisInQueue, nil
+}
+
+// newSDKProvisionedThroughtput builds a new *svcsdk.ProvisionedThroughput
```
Comment (jaypipes) [inline pkg/resource/table/hooks_global_secondary_indexes.go:209]:
```suggestion
// newSDKProvisionedThroughput builds a new *svcsdk.ProvisionedThroughput
```

File: pkg/resource/table/hooks_key_schema.go
Comment (jaypipes) [inline pkg/resource/table/hooks_key_schema.go:1]: I would combine this into a single `hooks.go` file. Not only does `newSDKKeySchemaArray` get defined in `hooks_global_secondary_indexes.go` but `newSDKAttributesDefinition` doesn't have anything to do with key schemas.

File: pkg/resource/table/hooks_local_secondary_indexes.go
Comment (jaypipes) [inline pkg/resource/table/hooks_local_secondary_indexes.go:1]: Combine into a single `hooks.go` file IMHO.

File: pkg/resource/table/hooks_local_secondary_indexes.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+package table
+
+import (
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackutil "github.com/aws-controllers-k8s/runtime/pkg/util"
+
+	"github.com/aws-controllers-k8s/dynamodb-controller/apis/v1alpha1"
+)
+
+func computeLocalSecondaryIndexDelta(
```
Comment (jaypipes) [inline pkg/resource/table/hooks_local_secondary_indexes.go:23]: Are local secondary indexes ever updated?

File: test/e2e/tag.py
Comment (jaypipes) [inline test/e2e/tag.py:1]: :+1:

File: test/e2e/tests/test_table.py [hunk 17]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+    def test_create_update_gsi(self, office_table):
+        (ref, res) = office_table
+
+        table_name = res["spec"]["tableName"]
+
+        # Check DynamoDB Table exists
+        assert self.table_exists(table_name)
+
+        # Get CR latest revision
+        cr = k8s.wait_resource_consumed_by_controller(ref)
```
Comment (jaypipes) [inline test/e2e/tests/test_table.py:657]: above tests are great. I don't see where local secondary indexes are ever tested though? or updated...

File: test/e2e/tests/test_global_table.py [hunk 18]
```diff
@@ -94,6 +94,7 @@ def get_global_table(self, dynamodb_client, global_table_name: str) -> dict:
     def global_table_exists(self, dynamodb_client, global_table_name: str) -> bool:
         return self.get_global_table(dynamodb_client, global_table_name) is not None
 
+    #@pytest.mark.skipif(True, reason="temporary")
```
Comment (jljaco) [inline test/e2e/tests/test_global_table.py:97]: Was this a leftover from testing?

File: test/e2e/tests/test_table.py [hunk 19]
```diff
@@ -67,15 +89,59 @@ def forum_table():
...
+    return table_reference, table_resource
+
+@pytest.fixture(scope="module")
+def forum_table():
+    resource_name = random_suffix_name("table-forums", 32)
+    (ref, cr) = create_table(resource_name, "table_forums")
+
+    yield ref, cr
+    try:
+        _, deleted = k8s.delete_custom_resource(ref, 3, 10)
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:105]: Nit: use named parameters (e.g., `period_length=X`)

File: test/e2e/tests/test_table.py [hunk 20]
```diff
@@ -104,22 +170,68 @@ def test_table_update_tags(self, forum_table):
...
+        tags.assert_ack_system_tags(
+            tags=table_tags,
+        )
+        tags_dict = tags.to_dict(
+            cr["spec"]["tags"],
+            key_member_name="key",
+            value_member_name="value"
+        )
+        tags.assert_equal_without_ack_tags(
+            actual=tags_dict,
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:195]: Nit: this should be reversed semantically (`actual=table_tags, expected=tags_dict`)

File: test/e2e/tests/test_table.py [hunk 21]
```diff
@@ -104,22 +170,68 @@ def test_table_update_tags(self, forum_table):
...
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+        time.sleep(MODIFY_WAIT_AFTER_SECONDS * 2)
+
+        table_tags = get_resource_tags(cr["status"]["ackResourceMetadata"]["arn"])
+        tags.assert_ack_system_tags(
+            tags=table_tags,
+        )
+        tags_dict = tags.to_dict(
+            cr["spec"]["tags"],
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:190]: As an added layer of validation, consider using `new_tags` here instead of `cr["spec"]["tags"]`

File: test/e2e/tests/test_table.py [hunk 22]
```diff
@@ -104,22 +170,68 @@ def test_table_update_tags(self, forum_table):
...
-        table_tags = tag.clean(get_resource_tags(cr["status"]["ackResourceMetadata"]["arn"]))
-        assert len(table_tags) == len(tags)
-        assert table_tags[0]['Key'] == tags[0]['key']
-        assert table_tags[0]['Value'] == tags[0]['value']
+        table_tags = get_resource_tags(cr["status"]["ackResourceMetadata"]["arn"])
+        tags.assert_ack_system_tags(
+            tags=table_tags,
+        )
+        tags_dict = tags.to_dict(
+            cr["spec"]["tags"],
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:227]: Ditto consider using `new_tags` here

File: test/e2e/tests/test_table.py [hunk 23]
```diff
@@ -104,22 +170,68 @@ def test_table_update_tags(self, forum_table):
...
+            key_member_name="key",
+            value_member_name="value"
+        )
+        tags.assert_equal_without_ack_tags(
+            actual=tags_dict,
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:232]: Ditto semantic reversal

File: test/e2e/tests/test_table.py [hunk 24]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+
+        table.wait_until(
+            table_name,
+            table.billing_mode_matcher("PROVISIONED"),
+            timeout_seconds=30,
+            interval_seconds=3,
+        )
+
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:344]: Add some `assert`ions here to validate that the values you set for read/write capacity are correctly set in `ProvisionedThroughput`

File: test/e2e/tests/test_table.py [hunk 25]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+            "streamEnabled": False,
+        }
+
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+
+        table.wait_until(
+            table_name,
+            table.stream_specification_matches(False),
+            timeout_seconds=30,
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:293]: Make these timeout and interval values global parameters @ top of file

File: test/e2e/tests/test_table.py [hunk 26]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+            }
+        }
+
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+
+        table.wait_until(
+            table_name,
+            table.billing_mode_matcher("PROVISIONED"),
+            timeout_seconds=30,
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:341]: Make these timeout and interval params global @ top of file

File: test/e2e/tests/test_table.py [hunk 27]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+
+        table.wait_until(
+            table_name,
+            table.billing_mode_matcher("PAY_PER_REQUEST"),
+            timeout_seconds=60*3,
+            interval_seconds=5,
+        ) """
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:362]: Add `assert`ions here to confirm that previously-set read/write capacities no longer present in `ProvisonedThroughput`

File: test/e2e/tests/test_table.py [hunk 28]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+            }
+        }
+
+        # Patch k8s resource
+        k8s.patch_custom_resource(ref, cr)
+
+        table.wait_until(
+            table_name,
+            table.sse_specification_matcher(True, "KMS"),
+            timeout_seconds=120,
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:418]: Make these timeout/interval param values global

File: test/e2e/tests/test_table.py [hunk 29]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+            interval_seconds=3,
+        )
+
+        time.sleep(60)
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:422]: Why is this necessary?  Make the value a global param w/ a descriptive name

File: test/e2e/tests/test_table.py [hunk 30]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+        table_name = res["spec"]["tableName"]
+
+        # Check DynamoDB Table exists
+        assert self.table_exists(table_name)
+
+        # Get CR latest revision
+        cr = k8s.wait_resource_consumed_by_controller(ref)
+
+        # Creating two more GSIs
+        cr["spec"] = {
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:514]: Do you mean to entirely overwrite `cr["spec"]` here?  why not just overwrite `cr["spec"]["attributeDefinitions"]` and `cr["spec"]["globalSecondaryIndexes"]`?

File: test/e2e/tests/test_table.py [hunk 31]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
+
+        # Check DynamoDB Table exists
+        assert self.table_exists(table_name)
+
+        # Get CR latest revision
+        cr = k8s.wait_resource_consumed_by_controller(ref)
+
+        # Creating two more GSIs
+        # Delete office-per-city and add office-per-country
+        cr["spec"] = {
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:573]: Ditto -- are you intending to completely overwrite the spec here?  Why not just update selectively?

File: test/e2e/tests/test_table.py [hunk 32]
```diff
@@ -155,3 +267,391 @@ def test_enable_ttl(self, forum_table):
...
```
Comment (jljaco) [inline test/e2e/tests/test_table.py:617]: Ditto spec overwrite question

File: pkg/resource/table/hooks_global_secondary_indexes.go (hunk 13 above)
Comment (A-Hilaly) [reply pkg/resource/table/hooks_global_secondary_indexes.go:79]: It is used in `customerPreCompare`:
```go
	if len(a.ko.Spec.GlobalSecondaryIndexes) != len(b.ko.Spec.GlobalSecondaryIndexes) {
		delta.Add("Spec.GlobalSecondaryIndexes", a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes)
	} else if a.ko.Spec.GlobalSecondaryIndexes != nil && b.ko.Spec.GlobalSecondaryIndexes != nil {
		if !equalGlobalSecondaryIndexesArrays(a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes) {
			delta.Add("Spec.GlobalSecondaryIndexes", a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes)
		}
	}
```

File: pkg/resource/table/hooks_local_secondary_indexes.go (hunk 16 above)
Comment (A-Hilaly) [reply pkg/resource/table/hooks_local_secondary_indexes.go:23]: Local secondary indexes can only be set during the CREATE op. They are immutable. `computeLocalSecondaryIndexDelta` is used to set a terminal condition if users try to modify them

File: test/e2e/tests/test_table.py (hunk 17 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_table.py:657]: Same reason here. LSIs are only mentioned during the creation, and the API will reject any non-valid LSIs, checking that the table is active is equivalent to checking the LSIs

File: test/e2e/tests/test_global_table.py (hunk 18 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_global_table.py:97]: Yep removed

File: test/e2e/tests/test_table.py (hunk 27 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_table.py:362]: The wait function already does that. It will fail if a field doesn't reach the intended state

File: test/e2e/tests/test_table.py (hunk 20 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_table.py:195]: Nice catch!

File: test/e2e/tests/test_table.py (hunk 29 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_table.py:422]: Because of tiny eventual consistency issues of DDB SSE Api

File: test/e2e/tests/test_table.py (hunk 30 above)
Comment (A-Hilaly) [reply test/e2e/tests/test_table.py:514]: This does a strategic merge, it doesn't remove the other fields :)
But i like your suggestion better, changing...

File: generator.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (A-Hilaly) [reply generator.yaml]: You're I brain lagged here. Addressed

**Review by jaypipes — APPROVED** [review]
Great work on this @A-Hilaly. I'm super excited to finally see this getting merged. :+1:

File: test/e2e/tests/test_table.py (hunk 24 above)
Comment (jljaco) [reply test/e2e/tests/test_table.py:341]: Are you sure?  [From my read](https://github.com/aws-controllers-k8s/dynamodb-controller/blob/117749ea0727440d258afc913a24e05f69c6432e/test/e2e/table.py#L97), `BillingModeMatcher` only matches on the `BillingModeSummary.BillingMode` field; there is no use of `ProvisionedThroughputMatcher` here.

File: test/e2e/tests/test_table.py (hunk 27 above)
Comment (jljaco) [reply test/e2e/tests/test_table.py:359]: See comment above
